src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Patterns.cs

```diff
@@ -1,272 +1,407 @@
+                if (pattern.Kind != BoundKind.RecursivePattern &&
+                    (_loweredInput.Kind == BoundKind.Local || _loweredInput.Kind == BoundKind.Parameter || _loweredInput.ConstantValue != null))
                 {
-                    var systemBoolean = _factory.SpecialType(SpecialType.System_Boolean);
-                    return RewriteNullableNullEquality(
-                        syntax: _factory.Syntax,
-                        kind: BinaryOperatorKind.NullableNullEqual,
-                        loweredLeft: input,
-                        loweredRight: boundConstant,
-                        returnType: systemBoolean);
+                    // Since non-recursive patterns cannot have side-effects on locals, we reuse an existing local
+                    // if present. A recursive pattern, on the other hand, may mutate a local through a captured lambda
+                    // when a `Deconstruct` method is invoked.
+                    _tempAllocator.AssignTemp(inputTemp, _loweredInput);
                 }
                 else
                 {
-                    return _factory.ObjectEqual(_factory.Convert(systemObject, input), boundConstant);
+                    // Even if subsequently unused (e.g. `GetValue() is _`), we assign to a temp to evaluate the side-effect
+                    _sideEffectBuilder.Add(_factory.AssignmentExpression(_tempAllocator.GetTemp(inputTemp), _loweredInput));
+                }
+
+                // then process all of the individual decisions
+                foreach (var decision in decisions)
+                {
+                    if (LowerDecision(decision))
+                    {
+                        break;
+                    }
+                }
+
+                if (_sideEffectBuilder.Count != 0)
+                {
+                    _conjunctBuilder.Add(_factory.Sequence(ImmutableArray<LocalSymbol>.Empty, _sideEffectBuilder.ToImmutable(), _factory.Literal(true)));
+                    _sideEffectBuilder.Clear();
+                }
+
+                BoundExpression result = null;
+                foreach (var conjunct in _conjunctBuilder)
+                {
+                    result = (result == null) ? conjunct : _factory.LogicalAnd(result, conjunct);
+                }
+
+                var bindingsBuilder = ArrayBuilder<BoundExpression>.GetInstance();
+                foreach (var (left, right) in bindings)
+                {
+                    bindingsBuilder.Add(_factory.AssignmentExpression(left, _tempAllocator.GetTemp(right)));
+                }
+
+                if (bindingsBuilder.Count > 0)
+                {
+                    var c = _factory.Sequence(ImmutableArray<LocalSymbol>.Empty, bindingsBuilder.ToImmutableAndFree(), _factory.Literal(true));
+                    result = (result == null) ? c : (BoundExpression)_factory.LogicalAnd(result, c);
                 }
+                else if (result == null)
```

Comment:
I believe this leaks an empty builder.

Reply:
Addressed in https://github.com/dotnet/roslyn/pull/23921/commits/f5d36158dca834088b681989f197f6b615d00284 of https://github.com/dotnet/roslyn/pull/23921